internal/controller/management/serviceaccounts/serviceaccounts_test.go

```diff
@@ -0,0 +1,117 @@
+package serviceaccounts
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func TestNewReconciler(t *testing.T) {
+	testCfg := ReconcilerConfig{
+		KargoNamespace: "fake-ns",
+	}
+	kubeClient := fake.NewClientBuilder().Build()
+	r := newReconciler(kubeClient, testCfg)
+	require.Equal(t, testCfg, r.cfg)
+	require.NotNil(t, kubeClient, r.client)
+
+	require.NotNil(t, r.getServiceAccountFn)
+	require.NotNil(t, r.updateServiceAccountFn)
+	require.NotNil(t, r.createRoleBindingFn)
+	require.NotNil(t, r.updateRoleBindingFn)
+	require.NotNil(t, r.deleteRoleBindingFn)
+	require.NotNil(t, r.listProjectFn)
+	require.NotNil(t, r.ensureControllerPermissionsFn)
+	require.NotNil(t, r.removeControllerPermissionsFn)
+}
+
+func TestEnsureControllerPermissions(t *testing.T) {
+	saName := "test-sa"
+	saNamespace := "default"
+	projectName := "project-1"
+
+	project := &kargoapi.Project{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+	serviceAccount := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName,
+			Namespace: saNamespace,
+		},
+	}
+
+	// Fake client with ServiceAccount and Project.
+	scheme := runtime.NewScheme()
+	_ = kargoapi.AddToScheme(scheme)
+	_ = corev1.AddToScheme(scheme)
+	_ = rbacv1.AddToScheme(scheme)
+	kubeClient := fake.NewClientBuilder().
+		WithScheme(scheme).
+		WithObjects(serviceAccount, project).
```

Comment:
Remember that this fake client is being built to help you simulate interactions with k8s APIs. The `ensureControllerPermissions()` method won't ever retrieve this SA or any SA. The SA is passed into it. i.e. You don't need to pass it to `WithObjects()` and you also won't need the corev1 scheme.